component/common/loki/client/queue_client.go

```diff
@@ -0,0 +1,584 @@
+		return nil, err
+	}
+
+	c.client.Timeout = cfg.Timeout
+
+	// Initialize counters to 0 so the metrics are exported before the first
+	// occurrence of incrementing to avoid missing metrics.
+	for _, counter := range c.metrics.countersWithHost {
+		counter.WithLabelValues(c.cfg.URL.Host).Add(0)
+	}
+
+	c.wg.Add(1)
+	go c.runSendOldBatches()
+	return c, nil
+}
+
+func (c *queueClient) initBatchMetrics(tenantID string) {
+	// Initialize counters to 0 so the metrics are exported before the first
+	// occurrence of incrementing to avoid missing metrics.
+	for _, counter := range c.metrics.countersWithHostTenantReason {
+		for _, reason := range Reasons {
+			counter.WithLabelValues(c.cfg.URL.Host, tenantID, reason).Add(0)
+		}
+	}
+
+	for _, counter := range c.metrics.countersWithHostTenant {
+		counter.WithLabelValues(c.cfg.URL.Host, tenantID).Add(0)
+	}
+}
+
+func (c *queueClient) SeriesReset(segmentNum int) {
+	c.seriesLock.Lock()
+	defer c.seriesLock.Unlock()
+	for k, v := range c.seriesSegment {
+		if v <= segmentNum {
+			level.Debug(c.logger).Log("msg", fmt.Sprintf("reclaiming series under segment %d", segmentNum))
+			delete(c.seriesSegment, k)
+			delete(c.series, k)
+		}
+	}
+}
+
+func (c *queueClient) StoreSeries(series []record.RefSeries, segment int) {
+	c.seriesLock.Lock()
+	defer c.seriesLock.Unlock()
+	for _, seriesRec := range series {
+		c.seriesSegment[seriesRec.Ref] = segment
+		labels := lokiutil.MapToModelLabelSet(seriesRec.Labels.Map())
+		c.series[seriesRec.Ref] = labels
+	}
+}
+
+func (c *queueClient) AppendEntries(entries wal.RefEntries, segment int) error {
+	c.seriesLock.RLock()
+	l, ok := c.series[entries.Ref]
+	c.seriesLock.RUnlock()
+	if ok {
+		for _, e := range entries.Entries {
+			c.appendSingleEntry(l, e)
+		}
+	} else {
+		// TODO(thepalbi): Add metric here
+		level.Debug(c.logger).Log("msg", "series for entry not found")
+	}
+	return nil
+}
+
+func (c *queueClient) appendSingleEntry(lbs model.LabelSet, e logproto.Entry) {
+	lbs, tenantID := c.processLabels(lbs)
+
+	// Either drop or mutate the log entry because its length is greater than maxLineSize. maxLineSize == 0 means disabled.
+	if c.maxLineSize != 0 && len(e.Line) > c.maxLineSize {
+		if !c.maxLineSizeTruncate {
+			c.metrics.droppedEntries.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonLineTooLong).Inc()
+			c.metrics.droppedBytes.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonLineTooLong).Add(float64(len(e.Line)))
+			return
+		}
+
+		c.metrics.mutatedEntries.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonLineTooLong).Inc()
+		c.metrics.mutatedBytes.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonLineTooLong).Add(float64(len(e.Line) - c.maxLineSize))
+		e.Line = e.Line[:c.maxLineSize]
+	}
+
+	// TODO: can I make this locking more fine grained?
+	c.batchesMtx.Lock()
+
+	batch, ok := c.batches[tenantID]
+
+	// If the batch doesn't exist yet, we create a new one with the entry
+	if !ok {
+		nb := newBatch(c.maxStreams)
+		_ = nb.addFromWAL(lbs, e)
```

Comment:
The assumption here is that a newly created batch should never error when adding a labelset? Maybe a comment about that?